### DISCUSSION TITLE
JanusGraph can parallel read properties by vid?

### DISCUSSION THREAD

**Discussion by wen313** [JanusGraph/janusgraph#2401] [Q&A] [answered]
for example：g.V().has('tag','janus').valueMap(true); if the tag of janus are many records,read properties by vid will be executed sequentially in backend,the time spent on network transmission will continue to accumulate..so can JanusGraph parallel read properties in the future?

**Comment by li-boxuan** [comment, marked as answer]
JanusGraph does support this. See _query.batch_ option in https://docs.janusgraph.org/basics/configuration-reference/#query but whether it takes effect depends on the data backend you are using. ~~For example, currently, Cassandra backend does not support it but HBase supports it~~. What is your data backend?

EDIT: just to update my answer to not mislead anyone who does not read the discussions below. In short, CQL backend supports "query.batch" option (see performance comparison between query.batch = false v.s. query.batch = true at https://www.experoinc.com/post/janusgraph-nuts-and-bolts-part-2-the-new-cql-adapter) but the option does not take effect on valueMap() step.

**Reply by wen313** [reply]
> JanusGraph does support this. See _query.batch_ option in https://docs.janusgraph.org/basics/configuration-reference/#query but whether it takes effect depends on the data backend you are using. For example, currently, Cassandra backend does not support it but HBase supports it. What is your data backend?

My data backend is Scylladb(Cassandra),So Can the Scylladb backend support this in the future?

**Reply by ryou90** [reply]
Since Scylla is a binary compatible variant of Cassandra and uses the same APIs, there will be no possibility there either.

**Reply by wen313** [reply]
So if I use Scylladb backend，Is there a way to use parallel read properties by vid？

**Reply by li-boxuan** [reply]
> So if I use Scylladb backend，Is there a way to use parallel read properties by vid？

At the moment you could use multi-threading at the application level.

Created an issue for this feature request: https://github.com/JanusGraph/janusgraph/issues/2406

**Reply by li-boxuan** [reply]
See my latest comment on https://github.com/JanusGraph/janusgraph/issues/2406#issuecomment-772543712. In short, using  _query.batch_ option in JanusGraph should satisfy your requirement.

**Reply by wen313** [reply]
but I find use `query.batch` is Invalid in valueMap query。

**Reply by li-boxuan** [reply]
@wen313 Yeah, I can confirm that optimization seems to be missing. It takes effect for values() but not valueMap(). Created https://github.com/JanusGraph/janusgraph/issues/2444
